File: apps/web/src/lib/client/__fixtures__/answerGaps.ts
/**
 * Answer-gap fixtures — Epic B.
 *
 * Shaped after the real `avp_dev` rows rather than invented, because the two
 * states that are easy to get wrong only show up together: `PSM Digital`'s
 * worst scan carries 9 genuine absences beside 10 prompts where no brand was
 * named at all, and a fixture without both cannot catch a screen that sums
 * them.
 */

import type { AnswerGaps, AnswerGapRow } from '@avp/shared-types';

const ENGINES = ['chatgpt', 'claude', 'claude_search'];

function row(
  promptId: string,
  text: string,
  kind: AnswerGapRow['kind'],
  {
    subjectNamedOn = 0,
    rivalNamedOn = 0,
    subjectCited = false,
    intent = 'awareness',
  }: {
    subjectNamedOn?: number;
    rivalNamedOn?: number;
    subjectCited?: boolean;
    intent?: string;
  } = {},
): AnswerGapRow {
  return {
    promptId,
    text,
    intent,
    position: 0,
    kind,
    enginesAnswered: 3,
    subjectNamedOn,
    rivalsNamedOn: rivalNamedOn,
    noBrandOn: kind === 'no_brands' ? 3 : 0,
    subjectCited,
    absentOn: 3 - subjectNamedOn,
    cells: [
      { brand: 'PSM Digital', namedOn: subjectNamedOn },
      { brand: 'WebFX', namedOn: rivalNamedOn },
      { brand: 'Clutch', namedOn: rivalNamedOn > 0 ? 1 : 0 },
    ],
  };
}

/** A gap-heavy scan: rivals winning, and a large no-brand tail beside them. */
export const gapHeavy: AnswerGaps = {
  clientId: 'clnt_01AAA',
  name: 'PSM Digital',
  domain: 'psmdigitalagency.com.au',
  scanId: 'scan_02',
  scannedAt: '2026-08-30T10:00:00Z',
  availableScanIds: ['scan_02', 'scan_01'],
  engines: ENGINES,
  brands: [
    { name: 'PSM Digital', domain: 'psmdigitalagency.com.au', isSubject: true, answersNamed: 4, promptsNamed: 2 },
    { name: 'WebFX', domain: 'webfx.com', isSubject: false, answersNamed: 9, promptsNamed: 3 },
    { name: 'Clutch', domain: 'clutch.co', isSubject: false, answersNamed: 8, promptsNamed: 3 },
  ],
  rows: [
    row('p1', 'how do i find a good seo agency in australia', 'absent', { rivalNamedOn: 3 }),
    row('p2', 'webfx vs a smaller australian seo agency', 'absent', { rivalNamedOn: 3 }),
    row('p3', 'what should i look for when hiring someone to run my google ads', 'no_brands'),
    row('p4', 'best local marketing help for a small shop', 'no_brands'),
    row('p5', 'psm digital reviews', 'covered', {
      subjectNamedOn: 3,
      rivalNamedOn: 1,
      subjectCited: true,
      intent: 'bottom_funnel',
    }),
    row('p6', 'affordable seo services for tradies', 'partial', {
      subjectNamedOn: 1,
      rivalNamedOn: 2,
      intent: 'comparison',
    }),
  ],
  rivals: [
    { name: 'WebFX', domain: 'webfx.com', answersWon: 9, scansPresent: 2 },
    { name: 'Clutch', domain: 'clutch.co', answersWon: 8, scansPresent: 1 },
  ],
  prompts: 6,
  absent: 2,
  partial: 1,
  uncited: 0,
  covered: 1,
  noBrands: 2,
  unanswered: 0,
  subjectCitable: true,
};

/**
 * A scan where the client's own domain was never cited.
 *
 * Modelled on the real Notion scan, where 22 prompts named the brand and none
 * cited its domain. `uncited` must stay 0 and the screen must say WHY rather
 * than showing a zero that reads as "no citation gaps".
 */
export const notCitable: AnswerGaps = {
  ...gapHeavy,
  name: 'Notion',
  domain: 'notion.so',
  subjectCitable: false,
  uncited: 0,
  absent: 0,
  covered: 4,
  noBrands: 0,
  partial: 2,
  prompts: 6,
  rivals: [],
  rows: gapHeavy.rows.map((r) => ({ ...r, kind: 'covered', subjectNamedOn: 3, absentOn: 0 })),
};

/** A scan with no rivals detected — the grid has one column and says so. */
export const noRivals: AnswerGaps = {
  ...gapHeavy,
  brands: [gapHeavy.brands[0]!],
  rows: gapHeavy.rows.map((r) => ({ ...r, cells: [r.cells[0]!] })),
  rivals: [],
};

/** One scan only, so the picker must not render. */
export const singleScan: AnswerGaps = {
  ...gapHeavy,
  availableScanIds: ['scan_02'],
};

File: apps/api/src/avp_api/schemas/answer_gaps.py
"""Answer gaps — where a rival is named and this client is not. Epic B.

WHAT A "GAP" IS HERE, AND THE THREE STATES THAT MUST NOT BE FOLDED TOGETHER
---------------------------------------------------------------------------
A prompt produced an answer from each engine. For this client, that answer is
in exactly one of three states, and they are different findings with different
fixes:

  * **named**      - the client appeared. No gap.
  * **absent**     - the client did not appear, but a rival did. THE gap: the
                     question has a commercial answer and somebody else is it.
  * **no brands**  - no brand appeared at all, from anyone. NOT a gap. The
                     question did not produce a brand answer, so there was
                     nothing to be absent from.

Folding `no brands` into `absent` would report a prompt nobody could have won
as a loss, and on the real `avp_dev` rows that is not a rounding error: one
scan carries 37 no-brand answers against 20 genuine absences. The same
discipline Epic A applied to `unclassified` sentiment, for the same reason.

WHY RECURRENCE IS COUNTED ACROSS ENGINES AND NOT ACROSS SCANS
-------------------------------------------------------------
The brief asked for gaps "sortable by how often the gap recurs", which assumed
prompts persist between scans. They do not. `build_prompt_set` calls
`generate_prompts` fresh for every scan, so prompt text is newly written each
time: across Plausible's three scans, 72 prompts carry 71 distinct texts, and
across Notion's two, 48 texts are 48 distinct. There is no prompt identity to
recur ON.

What does recur is the ENGINE. Every prompt is asked of every engine in the
scan, so a gap holding on 3 of 3 engines is a categorically stronger finding
than one holding on 1 of 3 — that is a real recurrence axis, it is measured
here as `absent_on` / `engines_answered`, and it is the default sort.

The other axis that survives between scans is the RIVAL, because competitors
are persistent rows. `AnswerGapRivalOut` counts those across the client's whole
history, which is the cross-scan half of the same question.

FACTS-ONLY RULE
--------------------------------
Every field below is a count, a boolean, an ordinal, an entity NAME or OUR OWN
prompt text. No engine answer text is read or returned — there is none stored
to read. `prompt.text` is ours: Epic 4 generated it.
"""

from __future__ import annotations

import enum
from datetime import datetime

from .common import ApiModel


class GapKind(str, enum.Enum):
    """One prompt's verdict for this client. Ordered by how bad it is.

    Precedence when several could apply is the declaration order below, worst
    first — a prompt that is both absent AND uncited is reported as `absent`,
    because being named at all comes before being cited.
    """

    # Every engine that answered named a rival and none named this client.
    ABSENT = "absent"
    # Some engines named this client and some did not. A gap with a foothold.
    PARTIAL = "partial"
    # Named by every engine that answered, but this client's own domain was
    # never cited for it — while the scan proves elsewhere that the domain IS
    # citable. Content exists; it is not the content answering this question.
    UNCITED = "uncited"
    # Named by every engine that answered.
    COVERED = "covered"
    # No brand at all was named, by anyone. Not a gap — see the module note.
    NO_BRANDS = "no_brands"
    # No engine produced a usable answer for this prompt.
    UNANSWERED = "unanswered"


class AnswerGapBrandOut(ApiModel):
    """One column of the grid: a brand, and how much of this scan it won."""

    name: str
    domain: str | None = None
    is_subject: bool
    # Answers (prompt x engine) that named this brand.
    answers_named: int
    # Prompts where at least one engine named it. The grid's column total.
    prompts_named: int


class AnswerGapCellOut(ApiModel):
    """One cell: how many of this prompt's engines named this brand.

    `named_on` is a count and not a boolean because the engines disagree, and
    that disagreement is the finding — a rival named by one engine of three is
    a different competitive position from one named by all three.
    """

    brand: str
    named_on: int


class AnswerGapRowOut(ApiModel):
    """One prompt, across every engine that answered it."""

    prompt_id: str
    # OURS — Epic 4 generated it. See the module note on the facts-only rule.
    text: str
    intent: str
    position: int

    engines_answered: int
    subject_named_on: int
    # Engines that named at least one rival. Distinguishes "a rival won this"
    # from "nobody was named", which `brands_named_on == 0` alone cannot.
    rivals_named_on: int
    # Engines that answered without naming ANY brand.
    no_brand_on: int

    # Was this client's own domain cited for this prompt, by any engine?
    subject_cited: bool

    kind: GapKind
    # Engines on which this client was absent. The recurrence measure, and the
    # default sort. Zero for a covered prompt.
    absent_on: int

    cells: list[AnswerGapCellOut]


class AnswerGapRivalOut(ApiModel):
    """A rival, counted across the client's WHOLE history.

    The cross-scan half of "how often does this recur". Competitors persist
    between scans where prompts do not, so this is the only axis on which a
    recurring gap can honestly be counted over time.
    """

    name: str
    domain: str | None = None
    # Answers where this rival was named and the client was not.
    answers_won: int
    # Distinct scans in which that happened at least once.
    scans_present: int


class AnswerGapsOut(ApiModel):
    """The Answer gaps read for one scan, plus the cross-scan rival rollup."""

    client_id: str
    name: str
    domain: str

    scan_id: str
    scanned_at: datetime
    # Every scan of this client that carries a grid, newest first, so the screen
    # can offer a picker without a second request.
    available_scan_ids: list[str]

    engines: list[str]
    brands: list[AnswerGapBrandOut]
    rows: list[AnswerGapRowOut]
    rivals: list[AnswerGapRivalOut]

    # --- totals, so the screen never has to re-derive them from `rows` ---
    prompts: int
    absent: int
    partial: int
    uncited: int
    covered: int
    no_brands: int
    unanswered: int
    # False when no answer in the scan cited this client's domain. When that is
    # so, `uncited` is NOT reported on any row: with no evidence the domain is
    # citable at all, "you have content but it wasn't cited" is a claim this
    # data cannot support. See `services/answer_gaps.py`.
    subject_citable: bool
